File: src/graphics/swapchain.zig
pub const std = @import("std");
pub const vk = @import("vk");
pub const glfw = @import("glfw");

const graphics = @import("../graphics.zig");
const Instance = graphics.Instance;
const Device = graphics.Device;
const Gpu = graphics.Gpu;

//

pub const Swapchain = struct {
    swapchain: vk.SwapchainKHR,
    format: vk.Format,
    extent: vk.Extent2D,
    images: []Image,
    suboptimal: bool,

    pub const Image = struct {
        image: vk.Image,
        view: vk.ImageView,
        index: u32,
    };

    const Self = @This();

    pub fn init(
        allocator: std.mem.Allocator,
        instance: Instance,
        gpu: *const Gpu,
        device: Device,
        surface: vk.SurfaceKHR,
        extent: vk.Extent2D,
    ) !Self {
        return try create(allocator, instance, gpu, device, surface, extent, .null_handle);
    }

    pub fn deinit(
        self: *const Self,
        allocator: std.mem.Allocator,
        device: Device,
    ) void {
        for (self.images) |*image| {
            device.destroyImageView(image.view, null);
        }
        allocator.free(self.images);

        device.destroySwapchainKHR(self.swapchain, null);
    }

    pub fn acquireImage(self: *Self, device: Device, on_acquire: vk.Semaphore) !*Image {
        while (true) {
            if (self.suboptimal) {
                // TODO: recreate swapchain
                // const old_swapchain = self.swapchain;
                // self.createSwapchain(old_swapchain);
                // self.
            }

            const result = try device.acquireNextImageKHR(self.swapchain, 1_000_000_000, on_acquire, .null_handle);
            if (result.result == .suboptimal_khr) {
                self.suboptimal = true;
            } else if (result.result == .timeout) {
                return error.SwapchainTimeout;
            } else if (result.result == .not_ready) {
                return error.SwapchainNotReady;
            }

            return &self.images[result.image_index];
        }
    }

    fn create(
        allocator: std.mem.Allocator,
        instance: Instance,
        gpu: *const Gpu,
        device: Device,
        surface: vk.SurfaceKHR,
        extent: vk.Extent2D,
        old_swapchain: vk.SwapchainKHR,
    ) !Self {
        const surface_formats = try instance.getPhysicalDeviceSurfaceFormatsAllocKHR(gpu.device, surface, allocator);
        const surface_format = try preferred_format(surface_formats);

        const present_modes = try instance.getPhysicalDeviceSurfacePresentModesAllocKHR(gpu.device, surface, allocator);
        const present_mode = try preferred_present_mode(present_modes);

        const caps = try instance.getPhysicalDeviceSurfaceCapabilitiesKHR(gpu.device, surface);

        var create_info = vk.SwapchainCreateInfoKHR{
            .surface = surface,
            .image_sharing_mode = undefined,
            .image_format = surface_format.format,
            .image_color_space = surface_format.color_space,
            .present_mode = present_mode,
            .image_extent = vk.Extent2D{
                .width = @max(@min(extent.width, caps.max_image_extent.width), caps.min_image_extent.width),
                .height = @max(@min(extent.height, caps.max_image_extent.height), caps.min_image_extent.height),
            },
            .min_image_count = 0,
            .image_array_layers = 1,
            .image_usage = .{
                // .color_attachment_bit = true, // render to a custom image and copy it over
                .transfer_dst_bit = true,
            },
            .pre_transform = .{ .identity_bit_khr = true },
            .composite_alpha = .{ .opaque_bit_khr = true },
            .clipped = vk.TRUE,
            .old_swapchain = old_swapchain,
        };

        create_info.min_image_count = caps.min_image_count + 1;
        if (caps.max_image_count != 0 and create_info.min_image_count > caps.max_image_count) {
            // max image count 0 means there is no max
            create_info.min_image_count = caps.max_image_count;
        }

        const queue_family_indices = .{ gpu.queue_families.graphics, gpu.queue_families.present };
        if (queue_family_indices[0] == queue_family_indices[1]) {
            create_info.image_sharing_mode = .exclusive;
            create_info.queue_family_index_count = 0;
            create_info.p_queue_family_indices = null;
        } else {
            create_info.image_sharing_mode = .concurrent;
            create_info.queue_family_index_count = 2;
            create_info.p_queue_family_indices = @ptrCast(&queue_family_indices);
        }

        const swapchain = try device.createSwapchainKHR(&create_info, null);
        errdefer device.destroySwapchainKHR(swapchain, null);
        // self.swapchain_format = create_info.image_format;
        // self.swapchain_extent = create_info.image_extent;
        // self.swapchain_suboptimal = false;

        const tmp_images = try device.getSwapchainImagesAllocKHR(swapchain, allocator);
        defer allocator.free(tmp_images);

        const images = try allocator.alloc(Image, tmp_images.len);
        errdefer allocator.free(images);

        for (tmp_images, images, 0..) |image, *saved_image, i| {
            saved_image.image = image;
            saved_image.index = @truncate(i);

            const image_view = device.createImageView(&vk.ImageViewCreateInfo{
                .image = image,
                .view_type = .@"2d",
                .format = create_info.image_format,
                .components = .{
                    .r = .identity,
                    .g = .identity,
                    .b = .identity,
                    .a = .identity,
                },
                .subresource_range = .{
                    .aspect_mask = .{ .color_bit = true },
                    .base_mip_level = 0,
                    .level_count = 1,
                    .base_array_layer = 0,
                    .layer_count = 1,
                },
            }, null);

            saved_image.view = image_view catch |err| {
                for (images, 0..i) |*deleting_view, _| {
                    device.destroyImageView(deleting_view.view, null);
                }
                return err;
            };
        }

        return Self{
            .swapchain = swapchain,
            .format = surface_format.format,
            .extent = extent,
            .images = images,
            .suboptimal = false,
        };
    }

    fn preferred_format(avail: []vk.SurfaceFormatKHR) !vk.SurfaceFormatKHR {
        if (avail.len == 0) {
            return error.NoSurfaceFormats;
        }

        // use B8G8R8A8_UNORM SRGB_NONLINEAR if its there
        for (avail) |avl| {
            if (avl.format == .b8g8r8a8_unorm and avl.color_space == .srgb_nonlinear_khr) {
                return avl;
            }
        }

        return avail[0];
    }

    fn preferred_present_mode(avail: []vk.PresentModeKHR) !vk.PresentModeKHR {
        // use MAILBOX if its there
        for (avail) |avl| {
            if (avl == .mailbox_khr) {
                return avl;
            }
        }

        // // fallback to IMMEDIATE
        // for (avail) |avl| {
        //     if (avl == .immediate_khr) {
        //         return avl;
        //     }
        // }

        return vk.PresentModeKHR.fifo_khr;
    }
};

File: src/graphics/swapchain.rs
use core::slice;
use std::sync::Arc;

use ash::{
    Device, Entry, Instance, khr,
    vk::{self, Handle},
};
use eyre::{Result, bail};
use winit::window::Window;

use crate::cold;

use super::queues::QueueFamilies;

//

pub struct Swapchain {
    window: Arc<Window>,

    inner: vk::SwapchainKHR,
    surface: vk::SurfaceKHR,
    gpu: vk::PhysicalDevice,
    pub extent: vk::Extent2D,
    format: vk::Format,
    images: Box<[vk::Image]>,
    suboptimal: bool,

    surface_loader: khr::surface::Instance,
    swapchain_loader: khr::swapchain::Device,
}

impl Swapchain {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        entry: &Entry,
        instance: &Instance,
        device: &Device,
        gpu: vk::PhysicalDevice,
        queue_families: &QueueFamilies,
        surface: vk::SurfaceKHR,
        extent: vk::Extent2D,
        window: Arc<Window>,
    ) -> Result<Self> {
        let surface_loader = khr::surface::Instance::new(entry, instance);
        let swapchain_loader = khr::swapchain::Device::new(instance, device);

        let res = Self::create(
            surface_loader,
            swapchain_loader,
            gpu,
            queue_families,
            surface,
            extent,
            window,
        )?;
        Ok(res)
    }

    pub fn recreate(&mut self, device: &Device, queue_families: &QueueFamilies) -> Result<()> {
        _ = unsafe { device.device_wait_idle() };

        let size = self.window.inner_size();
        let extent = vk::Extent2D {
            width: size.width,
            height: size.height,
        };

        self.destroy();

        *self = Self::create(
            self.surface_loader.clone(),
            self.swapchain_loader.clone(),
            self.gpu,
            queue_families,
            self.surface,
            extent,
            self.window.clone(),
        )?;

        Ok(())
    }

    pub fn acquire(
        &mut self,
        device: &Device,
        on_acquire: vk::Semaphore,
        queue_families: &QueueFamilies,
    ) -> Result<SwapchainImage> {
        loop {
            if self.suboptimal {
                self.recreate(device, queue_families)?;
            }

            let res = unsafe {
                self.swapchain_loader.acquire_next_image(
                    self.inner,
                    1_000_000_000, // 1 sec
                    on_acquire,
                    vk::Fence::null(),
                )
            };

            match res {
                Ok((index, suboptimal)) => {
                    self.suboptimal |= suboptimal;
                    return Ok(SwapchainImage {
                        image: self.images[index as usize],
                        index,
                    });
                }
                Err(vk::Result::NOT_READY) => continue,
                Err(vk::Result::TIMEOUT) => {
                    bail!("swapchain timeout")
                }
                Err(vk::Result::ERROR_OUT_OF_DATE_KHR) => {
                    self.recreate(device, queue_families)?;
                }
                Err(err) => {
                    eyre::bail!("failed to acquire next image: {err}")
                }
            }
        }
    }

    pub fn present(
        &mut self,
        image: SwapchainImage,
        queue: vk::Queue,
        wait_for: vk::Semaphore,
    ) -> Result<()> {
        let present_info = vk::PresentInfoKHR::default()
            .wait_semaphores(slice::from_ref(&wait_for))
            .swapchains(slice::from_ref(&self.inner))
            .image_indices(slice::from_ref(&image.index));
        self.suboptimal |= unsafe { self.swapchain_loader.queue_present(queue, &present_info)? };

        Ok(())
    }

    pub fn destroy(&mut self) {
        if self.inner.is_null() {
            cold();
            return;
        }

        unsafe { self.swapchain_loader.destroy_swapchain(self.inner, None) };
        self.inner = vk::SwapchainKHR::null();
    }

    fn create(
        surface_loader: khr::surface::Instance,
        swapchain_loader: khr::swapchain::Device,
        gpu: vk::PhysicalDevice,
        queue_families: &QueueFamilies,
        surface: vk::SurfaceKHR,
        extent: vk::Extent2D,
        window: Arc<Window>,
    ) -> Result<Self> {
        let surface_formats =
            unsafe { surface_loader.get_physical_device_surface_formats(gpu, surface)? };
        let surface_present_modes =
            unsafe { surface_loader.get_physical_device_surface_present_modes(gpu, surface)? };

        let surface_format = Self::preferred_format(&surface_formats);
        let present_mode = Self::preferred_present_mode(&surface_present_modes);

        let caps =
            unsafe { surface_loader.get_physical_device_surface_capabilities(gpu, surface)? };

        let mut image_count = caps.min_image_count + 1;
        if caps.max_image_count != 0 && image_count > caps.max_image_count {
            // max image count 0 means there is no max
            image_count = caps.max_image_count;
        }

        let queue_family_indices = [queue_families.present, queue_families.graphics];
        let (sharing_mode, queue_family_indices) =
            if queue_family_indices[0] == queue_family_indices[1] {
                (vk::SharingMode::EXCLUSIVE, &[][..])
            } else {
                (vk::SharingMode::CONCURRENT, &queue_family_indices[..])
            };

        let extent = vk::Extent2D {
            width: extent
                .width
                .max(caps.min_image_extent.width)
                .min(caps.max_image_extent.width),
            height: extent
                .height
                .max(caps.min_image_extent.height)
                .min(caps.max_image_extent.height),
        };

        let create_info = vk::SwapchainCreateInfoKHR::default()
            .surface(surface)
            .image_sharing_mode(sharing_mode)
            .queue_family_indices(queue_family_indices)
            .image_format(surface_format.format)
            .image_color_space(surface_format.color_space)
            .present_mode(present_mode)
            .image_extent(extent)
            .min_image_count(image_count)
            .image_array_layers(1)
            .image_usage(vk::ImageUsageFlags::TRANSFER_DST)
            .pre_transform(vk::SurfaceTransformFlagsKHR::IDENTITY)
            .composite_alpha(vk::CompositeAlphaFlagsKHR::OPAQUE)
            .clipped(true);

        let inner = unsafe { swapchain_loader.create_swapchain(&create_info, None)? };

        let images = unsafe { swapchain_loader.get_swapchain_images(inner)? }.into_boxed_slice();

        Ok(Self {
            window,

            inner,
            surface,
            gpu,
            extent,
            format: surface_format.format,
            images,
            suboptimal: false,

            surface_loader,
            swapchain_loader,
        })
    }

    fn preferred_format(formats: &[vk::SurfaceFormatKHR]) -> vk::SurfaceFormatKHR {
        formats
            .iter()
            .copied()
            .find(|f| {
                f.format == vk::Format::B8G8R8A8_UNORM
                    && f.color_space == vk::ColorSpaceKHR::SRGB_NONLINEAR
            })
            .unwrap_or(formats[0])
    }

    fn preferred_present_mode(present_modes: &[vk::PresentModeKHR]) -> vk::PresentModeKHR {
        present_modes
            .iter()
            .copied()
            .find(|mode| *mode == vk::PresentModeKHR::MAILBOX)
            .unwrap_or(vk::PresentModeKHR::FIFO)
    }
}

//

#[must_use]
#[derive(Debug)]
pub struct SwapchainImage {
    pub image: vk::Image,
    index: u32,
}
